addons/web_editor/static/src/js/editor/snippets.options.js

```diff
@@ -4183,23 +4183,56 @@ registry.BackgroundShape = SnippetOptionWidget.extend({
      * Returns the default colors for the currently selected shape.
      *
      * @private
-     */
-    _getDefaultColors() {
-        const $shapeContainer = this.$target.find('> .o_we_shape')
-            .clone()
-            .addClass('d-none')
-            // Needs to be in document for bg-image class to take effect
-            .appendTo(document.body);
-        const shapeContainer = $shapeContainer[0];
-        $shapeContainer.css('background-image', '');
-        const shapeSrc = shapeContainer && getBgImageURL(shapeContainer);
-        $shapeContainer.remove();
+     * @param {String} shape identifier of the shape to consider instead of the currenlty selected
+     */
+    _getDefaultColors(shape) {
+        let shapeSrc;
+        if (shape) {
+            // locate original in snippet options
+            const $shapeContainer = this.$el.find(".o_we_shape_menu we-button[data-shape='" + shape + "'] div.o_we_shape");
+            const shapeContainer = $shapeContainer[0];
+            shapeSrc = shapeContainer && getBgImageURL(shapeContainer);
+        } else {
+            const $shapeContainer = this.$target.find('> .o_we_shape')
+                .clone()
+                .addClass('d-none')
+                // Needs to be in document for bg-image class to take effect
+                .appendTo(document.body);
+            const shapeContainer = $shapeContainer[0];
+            $shapeContainer.css('background-image', '');
+            shapeSrc = shapeContainer && getBgImageURL(shapeContainer);
+            $shapeContainer.remove();
+        }
         if (!shapeSrc) {
             return {};
         }
         const url = new URL(shapeSrc, window.location.origin);
         return Object.fromEntries(url.searchParams.entries());
     },
+    /**
+     * Returns the implicit colors for the currently selected shape.
+     *
+     * The implicit colors are use upon shape selection. They are computed as:
+     * - the default colors
+     * - patched with each set of colors of previous siblings shape
+     * - patched with the colors of the previously selected shape
+     * - filtered to only keep the colors involved in the current shape
+     *
+     * @private
+     * @param {String} shape identifier of the selected shape
+     * @param {Object} previousColors colors of the shape before its replacement
+     */
+    _getImplicitColors(shape, previousColors) {
+        const defaultColors = this._getDefaultColors(shape);
+        let colors = _.defaults({}, previousColors);
+        let sibling = this.$target[0].previousElementSibling;
+        while (sibling) {
+            colors = _.defaults(colors, this._getShapeData(sibling).colors || {});
+            sibling = sibling.previousElementSibling;
+        }
+        colors = _.defaults(colors, defaultColors);
+        return _.pick(colors, _.keys(defaultColors));
```

Comment:
You can use `Object.keys` instead of `_.keys`.

`pick` can be done in a variety of ways, the one that I currently find most readable is:

```js
Object.fromEntries(keys.map(key => [key, obj[key]]));
// in your case
Object.fromEntries(Object.keys(defaultColors).map(key => [key, colors[key]]));
```

Not quite as readable as `_.pick` though, not sure if there are plans to completely get rid of underscorejs any time soon, maybe @qsm-odoo can weigh in on this one.